src/policies.rs

```diff
@@ -1,31 +1,124 @@
-    trace!("Executing policy: verify_git_commits");
+pub fn verify_git_commits<G: Git, P: Gpg>(config: &VerifyGitCommitsConfig, old_value: &str, new_value: &str,_ref_name: &str) -> Result<(), Box<dyn Error>> {
+    debug!("Executing policy: verify_git_commits");
+    
+    let git = G::new()?;
     let start = Instant::now();
+    let old_commit_id = Oid::from_str(old_value)?;
+    let new_commit_id = Oid::from_str(new_value)?;
+
+    let new_commit = git.find_commit(new_commit_id)?;
+    let mut commits = Vec::new();
+    if is_new_branch(old_value) {
+        info!("NEW BRANCH detected");
+        commits = git.find_unpushed_commits(new_commit_id)?;
+    } else if G::merge_commit(&new_commit)? {
+        info!("MERGE detected");
+        match new_commit.parents().nth(1) {
+            Some(second_parent) => {
+                commits.push(new_commit);
+                let common_ancestor_id = git.find_common_ancestor(old_commit_id, second_parent.id())?;
+                let mut commits2 = git.find_commits(common_ancestor_id, second_parent.id())?;
+                commits.append(&mut commits2);
+            },
+            None => return Err(Box::new(CapnError::new(format!("Second parent not found for merge commit"))))
+        };
+        verify_different_authors::<G>(&commits)?;
+    } else {
+        commits = git.find_commits(old_commit_id, new_commit_id)?;
+    }
+   
+    let commit_fingerprints = git.find_commit_fingerprints(&config.team_fingerprints_file, &commits)?;
+
+    if config.recv_keys_par {
+        let _result = P::par_receive_keys(&config.keyserver,&commit_fingerprints)?;
+    } else {
+        let _result = P::receive_keys(&config.keyserver,&commit_fingerprints)?;
+    }
 
-    let fingerprints = read_fingerprints::<G>(team_fingerprints_file)?;
-    let _result = P::par_receive_keys(&keyserver,&fingerprints)?;
+    verify_email_addresses(&config.author_domain, &config.committer_domain, &commits)?;
+
+    verify_commit_signatures::<G>(&git, &commits)?;
 
     let duration = start.elapsed();
-    trace!("verify_git_commits completed in: {}s", duration.as_secs());
+
+    info!("verify_git_commits completed in: {} ms", duration.as_millis());
+
+    Ok(())
+    // return Err(Box::new(CapnError::new(format!("Error on verify git commits for testing"))));
+}
+
+fn is_new_branch(from_id: &str) -> bool {
+    return from_id == DONT_CARE_REF;
+}
+
+fn verify_commit_signatures<G: Git>(git: &G, commits: &Vec<Commit<'_>>) -> Result<(), Box<dyn Error>> {
+    debug!("Verify commit signatures");
+    for commit in commits.iter() {
+        if G::not_merge_commit(commit)? {
+            let _fingerprint = git.verify_commit_signature(commit.id())?;
+        }
+    }
+    Ok(())
+}
+
+fn verify_different_authors<G: Git>(commits: &Vec<Commit<'_>>) -> Result<(), Box<dyn Error>> {
+    let mut authors = HashSet::new();
+    debug!("Verify different authors");
+    for commit in commits.iter() {
```

Comment:
Correct me if I'm wrong but I think that every time you create a mutable collection like this and insert into it in a loop you could instead map and still return from inside of the map.